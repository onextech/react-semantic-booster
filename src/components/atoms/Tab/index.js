import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Tab as SuiTab } from 'semantic-ui-react';
import { getCustomClassName, subtractObject, setStyledSpacer } from '../../../utils/helpers';


const basicClassName = 'basic';
const justifyClassName = 'justify';
const angularClassName = 'angular';
const secondaryClassName = 'secondary';
const subtleClassName = 'subtle';
const spacerClassName = 'spacer';

const StyledTab = styled(SuiTab)`
  &.${basicClassName} {
    .tab { padding: 0; }
    .segment { border: 0; }
    .ui.menu {
      border: 0;
      .item { margin: 0; }
    }
  }
  &.${justifyClassName} {
    .ui.menu {
      display: flex;
      justify-content: space-between;
      .item { flex-grow: 1; }
    }
  }
  &.${angularClassName} {
    .ui.menu .item { border-radius: 0; }
  }
  &.${secondaryClassName} {
    .ui.menu .item {
      color: grey;
      &.active { color: inherit; } 
    }
  }
  &.${subtleClassName} {
    .ui.menu .item:not(.active) {
      transition: color .75s;
      &:hover {
        background-color: inherit;
      }
    }
  }
  ${({ spacer }) => {
    if (spacer) {
      return `
        &.${spacerClassName} .ui.menu .item {
          ${setStyledSpacer(spacer, 1)}
        }
      `;
    }
    return false;
  }}
  `;

const Tab = (props) => {
  // 1. Define custom props for this component
  const classProps = {
    basic: basicClassName,
    justify: justifyClassName,
    angular: angularClassName,
    secondary: secondaryClassName,
    subtle: subtleClassName,
    spacer: spacerClassName,
  };
  // 1.5. Define default props
  const defaultProps = {
    basic: false,
    justify: false,
    angular: false,
    secondary: false,
    subtle: false,
    spacer: 0,
  };
  const customProps = {
    spacer: props.spacer,
  };

  const allProps = { ...defaultProps, ...props };
  // 2. Render the custom class names
  const className = props &&
    getCustomClassName(classProps, allProps);
  // 3. Clean up custom props from main prop set to avoid prop clashing
  const semanticProps = props &&
    subtractObject({ ...defaultProps, ...classProps }, allProps);

  return (
    <StyledTab {...semanticProps} {...customProps} className={className} />
  );
};

Tab.propTypes = {
  basic: PropTypes.bool,
  justify: PropTypes.bool,
  angular: PropTypes.bool,
  secondary: PropTypes.bool,
  subtle: PropTypes.bool,
  spacer: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      top: PropTypes.number,
      bottom: PropTypes.number,
    }),
  ]),
};

Tab.Pane = SuiTab.Pane;

export default Tab;

