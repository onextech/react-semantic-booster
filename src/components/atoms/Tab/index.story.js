import React from 'react';
import PropTypes from 'prop-types';
import { storiesOf } from '@storybook/react';
import Block from '../Block';
import Tab from '.';


const TabPane = ({ children }) => (
  <Tab.Pane>
    <Block secondary attached basic textAlign="center" spacer={2}>
      {children}
    </Block>
  </Tab.Pane>
);

TabPane.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string,
    PropTypes.array,
  ]).isRequired,
};

/* eslint-disable react/display-name */
const panes = [
  {
    menuItem: { key: 'users', icon: 'users', content: 'Users' },
    render: () => <TabPane>Tab 1 Content</TabPane>,
  },
  {
    menuItem: { key: '2', icon: 'wifi', content: 'Users 2' },
    render: () => <TabPane>Tab 2 Content</TabPane>,
  },
  {
    menuItem: { key: '3', icon: 'world', content: 'Globe' },
    render: () => <TabPane>Tab 3 Content</TabPane>,
  },
  {
    menuItem: { key: '4', icon: 'trash', content: 'Trash' },
    render: () => <TabPane>Tab 4 Content</TabPane>,
  },
];
/* eslint-enable react/display-name */

const menu = {
  icon: 'labeled',
  attached: true,
  secondary: true,
  tabular: true,
};

export const exampleTabProps = { // eslint-disable-line import/prefer-default-export
  spacer: 2,
  secondary: true,
  angular: true,
  justify: true,
  basic: true,
  subtle: true,
  panes,
  menu,
};

storiesOf('Tab', module)
  .add('Default', () => (
    <Tab {...exampleTabProps} />
  ));
