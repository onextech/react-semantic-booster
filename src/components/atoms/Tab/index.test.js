import React from 'react';
import renderer from 'react-test-renderer';
import Tab from '.';
import { exampleTabProps } from './index.story';


describe('<Tab>', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Tab {...exampleTabProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
