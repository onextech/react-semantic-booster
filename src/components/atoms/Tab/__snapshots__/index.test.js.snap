// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab> renders correctly 1`] = `
<div
  className="basic justify angular secondary subtle spacer   sc-bdVaJa edAoKK"
  spacer={2}
>
  <div
    className="ui secondary attached labeled icon tabular menu"
  >
    <a
      className="active item"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="users icon"
      />
      Users
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="wifi icon"
      />
      Users 2
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="world icon"
      />
      Globe
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="trash icon"
      />
      Trash
    </a>
  </div>
  <div
    className="ui bottom attached segment active tab"
  >
    <div
      className="ui basic secondary vertical attached center aligned segment spacer sc-bwzfXH bsANxE"
      spacer={2}
    >
      Tab 1 Content
    </div>
  </div>
</div>
`;
